AmplifyPlugins/Core/AWSPluginsCore/Model/Decorator/AuthRuleDecorator.swift

```diff
@@ -0,0 +1,106 @@
+//
+// Copyright 2018-2020 Amazon.com,
+// Inc. or its affiliates. All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import Amplify
+
+// https://github.com/aws-amplify/amplify-cli/issues/4182
+// If the owner authorization does not contain `.read` operation, then others can read owners data,
+// do not have to apply any owner field to the subscription document input
+/// Decorate the document with auth related fields such as the selection set and owner input to subscription documents
+public struct AuthRuleDecorator: ModelBasedGraphQLDocumentDecorator {
+
+    private let subscriptionType: GraphQLSubscriptionType?
+    private let ownerId: String?
+
+    public init(subscriptionType: GraphQLSubscriptionType? = nil, ownerId: String? = nil) {
+        self.subscriptionType = subscriptionType
+        self.ownerId = ownerId
+    }
+
+    public func decorate(_ document: SingleDirectiveGraphQLDocument,
+                         modelType: Model.Type) -> SingleDirectiveGraphQLDocument {
+        let authRules = modelType.schema.authRules
+        guard !authRules.isEmpty else {
+            return document
+        }
+        var decorateDocument = document
+        authRules.forEach { authRule in
+            decorateDocument = decorateIfOwnerAuthStrategy(document: decorateDocument, authRule: authRule)
+        }
+        return decorateDocument
+    }
+
+    func decorateIfOwnerAuthStrategy(document: SingleDirectiveGraphQLDocument,
+                                     authRule: AuthRule) -> SingleDirectiveGraphQLDocument {
+        guard authRule.allow == .owner else {
+            return document
+        }
+
+        guard var selectionSet = document.selectionSet else {
+            return document
+        }
+
+        let ownerField = authRule.getOwnerFieldOrDefault()
+        selectionSet = withOwnerField(selectionSet: selectionSet, ownerField: ownerField)
+
+        guard let subscriptionType = subscriptionType else {
+            return document.copy(selectionSet: selectionSet)
+        }
+
+        let operations = authRule.getModelOperationsOrDefault()
+
+        var requiredOwnerInput = false
+        switch subscriptionType {
+        case .onCreate:
+            if operations.contains(.create) {
+                requiredOwnerInput = true
+            }
+        case .onUpdate:
+            if operations.contains(.update) {
+                requiredOwnerInput = true
+            }
+        case .onDelete:
+            if operations.contains(.delete) {
+                requiredOwnerInput = true
+            }
+        }
+
+        if requiredOwnerInput, let ownerId = ownerId {
```

Comment:
I'm trying to think if we should error here if ownerId is `nil`.  Correct me if i'm wrong here, but I think the answer is "no": we should not have a warning here because if ownerId is nil, we are hitting a case where users have been logged out, and we should be able to still create a document without the ownerId and attempt to make the request, correct?

Reply:
this has got me thinking more and think it makes more sense that ownerId not optional in the AuthRuleDecorator because the decorators are supposed to be dumb and all it should do is to take the schema information and required inputs to decorate the document. This means it will only decorate the document input when 
1. we are in the context of providing a subscription document (input == .subscription)
2. ownerInput is required for this subscriptionType based on the auth rule's operation

By making ownerId required, the caller (DataStore) only has to check what type of backend is provisioned for that particular model. The backend is a conflict resolution enabled service by default, with a bunch of types with model directives. At runtime, it needs to check if the model.schema has authRules with owner auth strategy, then it will be responsible for checking if the user is signed in and retrieving the ownerId value. It will leave it up to the auth decorator, as part of building the GraphQLRequest, to know whether or not that ownerId is even needed.

Reply:
If the user has logged out and the backend is provisioned with authRules containing owner strategy, then the caller would be responsible for waiting for the user to sign back in to retrieve the ownerId. subscription graphQLRequests cannot be made, now that we have enforced ownerId to be non-optional. 

If user is signed in already, and the graphQLRequest is created for the subscription operation, then before Amplify.API.subscription is called, the user is signed out. the call to Amplify.APi.Subscription should then fail with unauthenticated since API will also try to intercept request header with the user pool token and fail

Reply:
made ownerId required